File: php/src/VedaResult.php
<?php

declare(strict_types=1);

namespace VedaDB;

/**
 * Represents a query result from VedaDB.
 */
class VedaResult
{
    /** @var list<string>|null */
    public ?array $columns;

    /** @var list<list<mixed>>|null */
    public ?array $rows;

    public int $rowCount;

    public ?string $message;

    /** @var array<string, mixed>|null */
    public ?array $metadata;

    public function __construct(
        ?array $columns,
        ?array $rows,
        int $rowCount,
        ?string $message,
        ?array $metadata = null
    ) {
        $this->columns  = $columns;
        $this->rows     = $rows;
        $this->rowCount = $rowCount;
        $this->message  = $message;
        $this->metadata = $metadata;
    }

    /**
     * Convert rows to associative arrays keyed by column name.
     *
     * @return list<array<string, mixed>>
     */
    public function toDicts(): array
    {
        if ($this->columns === null || $this->rows === null) {
            return [];
        }

        $result = [];
        foreach ($this->rows as $row) {
            $dict = [];
            foreach ($this->columns as $i => $col) {
                $dict[$col] = $row[$i] ?? null;
            }
            $result[] = $dict;
        }
        return $result;
    }

    /**
     * Get the first row as an associative array.
     *
     * @return array<string, mixed>|null
     */
    public function first(): ?array
    {
        $dicts = $this->toDicts();
        return $dicts[0] ?? null;
    }

    /**
     * Extract values from a single column.
     *
     * @return list<mixed>
     */
    public function pluck(string $column): array
    {
        if ($this->columns === null || $this->rows === null) {
            return [];
        }

        $idx = array_search($column, $this->columns, true);
        if ($idx === false) {
            return [];
        }

        return array_map(fn(array $row) => $row[$idx] ?? null, $this->rows);
    }

    /**
     * Get a column value from the first row.
     */
    public function value(string $column): mixed
    {
        $dict = $this->first();
        return $dict[$column] ?? null;
    }

    /**
     * Get all values from the first column.
     *
     * @return list<mixed>
     */
    public function column(): array
    {
        if ($this->rows === null || $this->columns === null) {
            return [];
        }
        return array_map(fn(array $row) => $row[0] ?? null, $this->rows);
    }

    /**
     * Check if the result is empty (no rows).
     */
    public function isEmpty(): bool
    {
        return $this->rowCount === 0;
    }

    /**
     * Return the number of rows.
     */
    public function count(): int
    {
        return $this->rowCount;
    }

    /**
     * Return a scalar value (single row, single column).
     */
    public function scalar(): mixed
    {
        if ($this->rows === null || empty($this->rows)) {
            return null;
        }
        return $this->rows[0][0] ?? null;
    }

    /**
     * Get a result set iterator.
     */
    public function getIterator(): VedaResultSet
    {
        return new VedaResultSet($this);
    }

    /**
     * Apply a callback to each row and return transformed results.
     *
     * @template T
     * @param callable(array<string, mixed>): T $callback
     * @return list<T>
     */
    public function map(callable $callback): array
    {
        return array_map($callback, $this->toDicts());
    }

    /**
     * Filter rows by a predicate.
     *
     * @param callable(array<string, mixed>): bool $predicate
     * @return list<array<string, mixed>>
     */
    public function filter(callable $predicate): array
    {
        return array_values(array_filter($this->toDicts(), $predicate));
    }

    /**
     * Parse a JSON response from VedaDB.
     */
    public static function parse(string $json): self
    {
        try {
            $data = json_decode($json, true, 512, JSON_THROW_ON_ERROR);
        } catch (\JsonException $e) {
            throw new VedaException('Failed to parse response: ' . $e->getMessage());
        }

        if (!is_array($data)) {
            throw new VedaException('Response is not a valid JSON object');
        }

        if (isset($data['error'])) {
            throw new QueryException(
                (string) $data['error'],
                isset($data['code']) ? (int) $data['code'] : null,
                $data['sqlstate'] ?? null,
            );
        }

        return new self(
            $data['columns'] ?? null,
            $data['rows'] ?? null,
            (int) ($data['row_count'] ?? 0),
            $data['message'] ?? null,
            $data['metadata'] ?? null,
        );
    }
}

File: dotnet/VedaResult.cs
using System.Text.Json;
using System.Text.Json.Serialization;

namespace VedaDB;

/// <summary>
/// Represents the result of a VedaDB query.
/// </summary>
public class VedaResult
{
    [JsonPropertyName("columns")]
    public List<string>? Columns { get; set; }

    [JsonPropertyName("rows")]
    public List<List<JsonElement>>? Rows { get; set; }

    [JsonPropertyName("row_count")]
    public int RowCount { get; set; }

    [JsonPropertyName("message")]
    public string? Message { get; set; }

    [JsonPropertyName("error")]
    public string? Error { get; set; }

    /// <summary>
    /// Convert rows to a list of dictionaries keyed by column name.
    /// </summary>
    public List<Dictionary<string, string?>> ToDicts()
    {
        var result = new List<Dictionary<string, string?>>();
        if (Columns == null || Rows == null) return result;

        foreach (var row in Rows)
        {
            var dict = new Dictionary<string, string?>();
            for (int i = 0; i < Columns.Count && i < row.Count; i++)
            {
                dict[Columns[i]] = row[i].ValueKind == JsonValueKind.Null
                    ? null
                    : row[i].ToString();
            }
            result.Add(dict);
        }
        return result;
    }

    /// <summary>
    /// Get the first row as a dictionary, or null if empty.
    /// </summary>
    public Dictionary<string, string?>? First()
    {
        var dicts = ToDicts();
        return dicts.Count > 0 ? dicts[0] : null;
    }

    /// <summary>
    /// Extract a single column's values as a list.
    /// </summary>
    public List<string?> Pluck(string column)
    {
        var result = new List<string?>();
        if (Columns == null || Rows == null) return result;

        int idx = Columns.IndexOf(column);
        if (idx < 0) return result;

        foreach (var row in Rows)
        {
            if (idx < row.Count)
            {
                result.Add(row[idx].ValueKind == JsonValueKind.Null
                    ? null
                    : row[idx].ToString());
            }
        }
        return result;
    }

    /// <summary>
    /// Get a single scalar value from the first row, first column.
    /// </summary>
    public string? Scalar()
    {
        if (Rows == null || Rows.Count == 0) return null;
        if (Rows[0].Count == 0) return null;
        return Rows[0][0].ValueKind == JsonValueKind.Null
            ? null
            : Rows[0][0].ToString();
    }

    /// <summary>
    /// Check if the result has any rows.
    /// </summary>
    public bool HasRows => Rows != null && Rows.Count > 0;

    /// <summary>
    /// Check if the result is empty (no rows).
    /// </summary>
    public bool IsEmpty => !HasRows;

    /// <summary>
    /// Get rows as a list of strongly-typed objects using a mapper function.
    /// </summary>
    public List<T> Map<T>(Func<Dictionary<string, string?>, T> mapper)
    {
        return ToDicts().Select(mapper).ToList();
    }

    /// <summary>
    /// Get the first row mapped to a strongly-typed object, or default if empty.
    /// </summary>
    public T? MapFirst<T>(Func<Dictionary<string, string?>, T> mapper)
    {
        var dicts = ToDicts();
        return dicts.Count > 0 ? mapper(dicts[0]) : default;
    }

    internal static VedaResult Parse(string json)
    {
        var result = JsonSerializer.Deserialize<VedaResult>(json)
            ?? throw new VedaException("Failed to parse response");

        if (!string.IsNullOrEmpty(result.Error))
        {
            throw new VedaQueryException(result.Error);
        }

        return result;
    }
}
